helix-core/src/main/java/org/apache/helix/common/caches/ParticipantStateCache.java

```diff
@@ -0,0 +1,177 @@
+          instanceMap = Maps.newHashMap();
+          allParticipantStateMap.put(instanceName, instanceMap);
+        }
+        Map<String, T> customizedMap = instanceMap.get(customizedName);
+        if (customizedMap == null) {
+          customizedMap = Maps.newHashMap();
+          instanceMap.put(customizedName, customizedMap);
+        }
+        customizedMap.put(resourceName, participantState);
+      } else {
+        LogUtil.logError(LOG, genEventInfo(), "Invalid key found in the participant state cache" + key);
+      }
+    }
+
+    _participantStateMap = Collections.unmodifiableMap(allParticipantStateMap);
+
+    long endTime = System.currentTimeMillis();
+    LogUtil.logInfo(LOG, genEventInfo(),
+        "END: participantStateCache.refresh() for cluster " + _controlContextProvider.getClusterName()
+            + ", started at : " + startTime + ", took " + (endTime - startTime) + " ms");
+    if (LOG.isDebugEnabled()) {
+      LogUtil.logDebug(LOG, genEventInfo(),
+          String.format("Participant State refreshed : %s", _participantStateMap.toString()));
+    }
+    return true;
+  }
+
+  // reload participant states that has been changed from zk to local cache.
+  private void refreshParticipantStatesCacheFromZk(HelixDataAccessor accessor,
+      Map<String, LiveInstance> liveInstanceMap, Boolean snapshotEnabled) {
+
+    long start = System.currentTimeMillis();
+    Set<PropertyKey> participantStateKeys = PopulateParticipantKeys(accessor, liveInstanceMap);
+
+    // All new entries from zk not cached locally yet should be read from ZK.
+    Set<PropertyKey> reloadKeys = new HashSet<>(participantStateKeys);
+    reloadKeys.removeAll(_participantStateCache.keySet());
+
+    Set<PropertyKey> cachedKeys = new HashSet<>(_participantStateCache.keySet());
+    cachedKeys.retainAll(participantStateKeys);
+
+    Set<PropertyKey> reloadedKeys = new HashSet<>();
+    Map<PropertyKey, T> newStateCache = Collections.unmodifiableMap(
+        refreshProperties(accessor, reloadKeys, new ArrayList<>(cachedKeys), _participantStateCache, reloadedKeys));
+
+    if (snapshotEnabled) {
+      refreshSnapshot(newStateCache, _participantStateCache, reloadedKeys);
+    }
+
+    _participantStateCache = newStateCache;
+
+    if (LOG.isDebugEnabled()) {
+      LogUtil.logDebug(LOG, genEventInfo(),
+          "# of participant state reload: " + reloadKeys.size() + ", skipped:" + (participantStateKeys.size()
+              - reloadKeys.size()) + ". took " + (System.currentTimeMillis() - start)
+              + " ms to reload new participant states for cluster: " + _controlContextProvider.getClusterName()
+              + "and state: " + this.getClass().getName());
+    }
+  }
+
+  protected abstract Set<PropertyKey> PopulateParticipantKeys(HelixDataAccessor accessor,
+      Map<String, LiveInstance> liveInstanceMap);
+
+  protected abstract void refreshSnapshot(Map<PropertyKey, T> newStateCache, Map<PropertyKey, T> participantStateCache,
```

Comment:
Since the customized state cache won't support snapshot for now, just comment on this method as OPTIONAL to be implemented. If the child class does not support snapshot, just do nothing. And we don't need the boolean to differentiate.

We might want to support it later though.